icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -740,10 +740,17 @@ int32_t MeasureUnit::getDimensionality(UErrorCode& status) const {
     return SingleUnitImpl::forMeasureUnit(*this, status).dimensionality;
 }
 
-MeasureUnit MeasureUnit::withDimensionality(int32_t dimensionality, UErrorCode& status) const {
-    SingleUnitImpl singleUnit = SingleUnitImpl::forMeasureUnit(*this, status);
-    singleUnit.dimensionality = dimensionality;
-    return singleUnit.build(status);
+MeasureUnit MeasureUnit::withDimensionality(int32_t dimensionality, UErrorCode &status) const {
+    MeasureUnit result;
+    int32_t unitsCount;
+    auto singleUnits = this->splitToSingleUnits(unitsCount, status);
+    for (int i = 0; i < unitsCount; i++) {
+        SingleUnitImpl singleUnit = SingleUnitImpl::forMeasureUnit(singleUnits[i], status);
+        singleUnit.dimensionality *= dimensionality;
```

Comment:
This PR is a behaviour change.

withDimensionality was idempotent: kilometer.withDimensionality(2).withDimensionality(2) should end up having dimensionality 2 - based on how with* is used in fluent APIs.

I'm not sure what a better name would be. Some thoughts:
- "meter.square().square()" to get p4-meter
- "meter.power(2)" ...

I don't think this makes sense for MeasureUnit though: we're not doing symbolic maths, we're declaring Measurement types. I think this code might be a better fit with the unit conversion code, not the MeasureUnit code: we need code like this for conversions, whereas for any other kind of MeasureUnit usage, I can't think of a use case that would really make sense?

Reply:
As I understand, MeasureUnit#withDimentionality means that a specific unit is going to take another dimensionality.

For example: if I have  `meter-per-second`  as `MeasureUnit`. And I would like the same unit with dimensionality = 2. Then, it will be `square-meter-per-square-second`

Reply:
withDimensionality is paired with getDimensionality. getDimensionality is meaningless for COMPOUND units, e.g. `meter-per-second-squared`. It's a concept that applies to single units only. And as mentioned withDimensionality is used to set the dimensionality to a specific value.

I think withDimensionality also doesn't make sense with MIXED units (foot-and-inch) - in that I don't see any use case where one would want to convert `foot-and-inch` to `square-foot-and-square-inch`: our only use case (at the moment) for mixed units is as a set of output preferences for a locale.

Hence my suggestion if you want a "raise this measure unit to the power of N" function, which would also take `meter-per-square-second` to `square-meter-per-p4-second` (something different than "set dimensionality to 2", it should be called something else.

I suggest letting the "raise to the power N" code live in unitconverter.cpp or similar.